File: rust/src/robot.rs
use crate::bridge::get_robot_registry_client;
use crate::core::types::{Metadata, RobotType, SensorType};
use crate::dataviz::DataViz;
use crate::sensors::{Sensor, SensorRef};
use crate::utils::{get_topic_monitor, get_topic_status_board};
use serde_json::{json, Value};
use std::sync::Arc;

#[derive(Debug, Clone)]
pub struct RobotManagerConfig {
    pub enabled: bool,
    pub status: bool,
    pub data_viz: bool,
    pub teleop: bool,
    pub tasks: bool,
}

impl RobotManagerConfig {
    pub fn enabled() -> Self {
        Self::default()
    }
}

impl Default for RobotManagerConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            status: true,
            data_viz: true,
            teleop: true,
            tasks: true,
        }
    }
}

#[derive(Debug, Clone)]
pub struct TeleopConfig {
    pub enabled: bool,
    pub command_topic: Option<String>,
    pub raw_input_topic: Option<String>,
    pub head_pose_topic: Option<String>,
    pub robot_profile: Option<String>,
    pub response_mode: Option<String>,
    pub publish_rate_hz: Option<f64>,
    pub custom_passthrough_only: Option<bool>,
    pub deadman_policy: Option<String>,
    pub deadman_timeout_ms: Option<i64>,
    pub deadzone: Option<f64>,
    pub expo: Option<f64>,
    pub linear_xy_max_mps: Option<f64>,
    pub linear_z_max_mps: Option<f64>,
    pub angular_z_max_rps: Option<f64>,
    pub invert_linear_x: Option<bool>,
    pub invert_linear_y: Option<bool>,
    pub invert_linear_z: Option<bool>,
    pub invert_angular_z: Option<bool>,
    pub discrete_threshold: Option<f64>,
    pub linear_xy_step_mps: Option<f64>,
    pub linear_z_step_mps: Option<f64>,
    pub angular_z_step_rps: Option<f64>,
}

impl TeleopConfig {
    pub fn enabled() -> Self {
        Self::default()
    }
}

impl Default for TeleopConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            command_topic: None,
            raw_input_topic: None,
            head_pose_topic: None,
            robot_profile: None,
            response_mode: None,
            publish_rate_hz: None,
            custom_passthrough_only: None,
            deadman_policy: None,
            deadman_timeout_ms: None,
            deadzone: None,
            expo: None,
            linear_xy_max_mps: None,
            linear_z_max_mps: None,
            angular_z_max_rps: None,
            invert_linear_x: None,
            invert_linear_y: None,
            invert_linear_z: None,
            invert_angular_z: None,
            discrete_threshold: None,
            linear_xy_step_mps: None,
            linear_z_step_mps: None,
            angular_z_step_rps: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct NavigationTaskConfig {
    pub go_to_point_enabled: bool,
    pub waypoint_enabled: bool,
    pub goal_topic: Option<String>,
    pub cancel_topic: Option<String>,
    pub goal_status_topic: Option<String>,
    pub waypoint_path_topic: Option<String>,
    pub waypoint_status_topic: Option<String>,
    pub frame_id: String,
    pub position_tolerance_m: Option<f64>,
    pub yaw_tolerance_deg: Option<f64>,
    pub min_altitude_m: Option<f64>,
    pub max_altitude_m: Option<f64>,
}

impl Default for NavigationTaskConfig {
    fn default() -> Self {
        Self {
            go_to_point_enabled: true,
            waypoint_enabled: true,
            goal_topic: None,
            cancel_topic: None,
            goal_status_topic: None,
            waypoint_path_topic: None,
            waypoint_status_topic: None,
            frame_id: "map".to_string(),
            position_tolerance_m: None,
            yaw_tolerance_deg: None,
            min_altitude_m: None,
            max_altitude_m: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct RobotDescriptionConfig {
    pub urdf_path: String,
    pub base_frame: String,
    pub source: String,
    pub ros_param_node: String,
    pub ros_param_name: String,
    pub chunk_size_bytes: i64,
    pub is_transparent: bool,
    pub include_visual_meshes: bool,
    pub visual_mesh_triangle_budget: i64,
    pub body_mesh_mode: String,
    pub enabled: bool,
}

impl RobotDescriptionConfig {
    pub fn new(urdf_path: impl Into<String>, base_frame: impl Into<String>) -> Self {
        Self {
            urdf_path: urdf_path.into(),
            base_frame: base_frame.into(),
            source: "ros".to_string(),
            ros_param_node: String::new(),
            ros_param_name: "robot_description".to_string(),
            chunk_size_bytes: 12000,
            is_transparent: false,
            include_visual_meshes: true,
            visual_mesh_triangle_budget: 90000,
            body_mesh_mode: "preview_mesh".to_string(),
            enabled: true,
        }
    }
}

#[derive(Debug, Clone)]
pub struct RobotDimensions {
    pub length: f64,
    pub width: f64,
    pub height: f64,
}

impl RobotDimensions {
    pub fn new(length: f64, width: f64, height: f64) -> Self {
        assert!(length >= 0.0 && width >= 0.0 && height >= 0.0);
        Self {
            length,
            width,
            height,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Robot {
    pub name: String,
    pub robot_type: RobotType,
    pub metadata: Metadata,
    pub sensors: Vec<SensorRef>,
    pub dimensions: Option<RobotDimensions>,
}

impl Robot {
    pub fn new(name: impl Into<String>, robot_type: RobotType) -> Self {
        let name = name.into();
        assert!(!name.is_empty(), "Robot name cannot be empty");
        Self {
            name,
            robot_type,
            metadata: Metadata::new(),
            sensors: Vec::new(),
            dimensions: None,
        }
    }

    pub fn with_dimensions(
        name: impl Into<String>,
        robot_type: RobotType,
        dimensions: RobotDimensions,
    ) -> Self {
        let mut robot = Self::new(name, robot_type);
        robot.dimensions = Some(dimensions);
        robot
    }

    pub fn get_type_str(&self) -> &'static str {
        self.robot_type.as_str()
    }

    pub fn add_metadata(&mut self, key: impl Into<String>, value: Value) {
        self.metadata.insert(key.into(), value);
    }

    pub fn get_metadata(&self, key: &str, default: Option<Value>) -> Option<Value> {
        self.metadata.get(key).cloned().or(default)
    }

    pub fn set_metadata(&mut self, key: impl Into<String>, value: Value) {
        self.add_metadata(key, value);
    }

    pub fn configure_ros_binding(
        &mut self,
        tf_mode: &str,
        topic_mode: &str,
        base_frame: Option<&str>,
    ) {
        let tf_mode = normalize_binding_mode(tf_mode);
        let topic_mode = normalize_binding_mode(topic_mode);
        let base_frame = base_frame
            .map(str::trim)
            .filter(|value| !value.is_empty())
            .unwrap_or("base_link");
        let tf_prefix = if tf_mode == "flat" { "" } else { self.name.as_str() };
        let topic_prefix = if topic_mode == "flat" {
            String::new()
        } else {
            format!("/{}", self.name)
        };
        self.metadata.insert(
            "ros_binding".to_string(),
            json!({
                "logical_name": self.name,
                "tf_mode": tf_mode,
                "topic_mode": topic_mode,
                "base_frame": base_frame,
                "tf_prefix": tf_prefix,
                "topic_prefix": topic_prefix,
            }),
        );
    }

    pub fn get_ros_binding(&self) -> Value {
        self.metadata
            .get("ros_binding")
            .and_then(Value::as_object)
            .cloned()
            .map(Value::Object)
            .unwrap_or_else(|| {
                json!({
                    "logical_name": self.name,
                    "tf_mode": "prefixed",
                    "topic_mode": "prefixed",
                    "base_frame": "base_link",
                    "tf_prefix": self.name,
                    "topic_prefix": format!("/{}", self.name),
                })
            })
    }

    pub fn resolve_topic(&self, topic_name: &str) -> String {
        let topic = topic_name.trim();
        if topic.is_empty() {
            return String::new();
        }
        if topic.starts_with('/') {
            return topic.to_string();
        }
        let binding = self.get_ros_binding();
        let prefix = binding
            .get("topic_prefix")
            .and_then(Value::as_str)
            .unwrap_or("");
        if prefix.trim().is_empty() {
            format!("/{topic}")
        } else {
            format!("{}/{}", prefix.trim_end_matches('/'), topic)
        }
    }

    pub fn resolve_tf_frame(&self, frame_id: &str) -> String {
        let frame = frame_id.trim();
        if frame.is_empty() {
            return String::new();
        }
        if frame.starts_with('/') {
            return frame.trim_start_matches('/').to_string();
        }
        let binding = self.get_ros_binding();
        let prefix = binding
            .get("tf_prefix")
            .and_then(Value::as_str)
            .unwrap_or("");
        if prefix.trim().is_empty() || frame.starts_with(&format!("{}/", prefix)) {
            frame.to_string()
        } else {
            format!("{}/{}", prefix.trim_matches('/'), frame)
        }
    }

    pub fn configure_robot_manager(&mut self, status: bool, data_viz: bool, teleop: bool, tasks: bool) {
        self.configure_robot_manager_with(RobotManagerConfig {
            enabled: true,
            status,
            data_viz,
            teleop,
            tasks,
        });
    }

    pub fn configure_robot_manager_with(&mut self, config: RobotManagerConfig) {
        self.metadata.insert(
            "robot_manager_config".to_string(),
            json!({
                "enabled": config.enabled,
                "prefab_asset_path": "Assets/Prefabs/UI/RobotManager.prefab",
                "prefab_resource_path": "",
                "sections": {
                    "status": config.status,
                    "data_viz": config.data_viz,
                    "teleop": config.teleop,
                    "tasks": config.tasks,
                }
            }),
        );
    }

    pub fn configure_teleop(&mut self, config: TeleopConfig) {
        let mut payload = serde_json::Map::new();
        payload.insert("enabled".to_string(), json!(config.enabled));
        put_if_some(&mut payload, "command_topic", config.command_topic);
        put_if_some(&mut payload, "raw_input_topic", config.raw_input_topic);
        put_if_some(&mut payload, "head_pose_topic", config.head_pose_topic);
        put_if_some(&mut payload, "robot_profile", config.robot_profile);
        put_if_some(&mut payload, "response_mode", config.response_mode);
        put_if_some(&mut payload, "publish_rate_hz", config.publish_rate_hz);
        put_if_some(&mut payload, "custom_passthrough_only", config.custom_passthrough_only);

        let mut deadman = serde_json::Map::new();
        put_if_some(&mut deadman, "policy", config.deadman_policy);
        put_if_some(&mut deadman, "timeout_ms", config.deadman_timeout_ms);
        if !deadman.is_empty() {
            payload.insert("deadman".to_string(), Value::Object(deadman));
        }

        let mut axes = serde_json::Map::new();
        put_if_some(&mut axes, "deadzone", config.deadzone);
        put_if_some(&mut axes, "expo", config.expo);
        put_if_some(&mut axes, "linear_xy_max_mps", config.linear_xy_max_mps);
        put_if_some(&mut axes, "linear_z_max_mps", config.linear_z_max_mps);
        put_if_some(&mut axes, "angular_z_max_rps", config.angular_z_max_rps);
        put_if_some(&mut axes, "invert_linear_x", config.invert_linear_x);
        put_if_some(&mut axes, "invert_linear_y", config.invert_linear_y);
        put_if_some(&mut axes, "invert_linear_z", config.invert_linear_z);
        put_if_some(&mut axes, "invert_angular_z", config.invert_angular_z);
        if !axes.is_empty() {
            payload.insert("axes".to_string(), Value::Object(axes));
        }

        let mut discrete = serde_json::Map::new();
        put_if_some(&mut discrete, "threshold", config.discrete_threshold);
        put_if_some(&mut discrete, "linear_xy_step_mps", config.linear_xy_step_mps);
        put_if_some(&mut discrete, "linear_z_step_mps", config.linear_z_step_mps);
        put_if_some(&mut discrete, "angular_z_step_rps", config.angular_z_step_rps);
        if !discrete.is_empty() {
            payload.insert("discrete".to_string(), Value::Object(discrete));
        }

        self.metadata.insert("teleop_config".to_string(), Value::Object(payload));
    }

    pub fn configure_navigation_tasks(&mut self, config: NavigationTaskConfig) {
        self.metadata.insert(
            "task_config".to_string(),
            json!({
                "go_to_point": {
                    "enabled": config.go_to_point_enabled,
                    "goal_topic": config.goal_topic,
                    "cancel_topic": config.cancel_topic,
                    "status_topic": config.goal_status_topic,
                    "frame_id": config.frame_id,
                    "position_tolerance_m": config.position_tolerance_m,
                    "yaw_tolerance_deg": config.yaw_tolerance_deg,
                    "min_altitude_m": config.min_altitude_m,
                    "max_altitude_m": config.max_altitude_m,
                },
                "waypoint": {
                    "enabled": config.waypoint_enabled,
                    "path_topic": config.waypoint_path_topic,
                    "status_topic": config.waypoint_status_topic,
                    "frame_id": config.frame_id,
                    "position_tolerance_m": config.position_tolerance_m,
                    "yaw_tolerance_deg": config.yaw_tolerance_deg,
                }
            }),
        );
    }

    pub fn configure_robot_description(&mut self, config: RobotDescriptionConfig) {
        let mut body_mesh_mode = config.body_mesh_mode.trim().to_ascii_lowercase();
        if body_mesh_mode == "max_quality_mesh" {
            body_mesh_mode = "runtime_high_mesh".to_string();
        }
        if !["collision_only", "preview_mesh", "runtime_high_mesh"].contains(&body_mesh_mode.as_str()) {
            body_mesh_mode = "preview_mesh".to_string();
        }
        let include_visual_meshes = config.include_visual_meshes && body_mesh_mode != "collision_only";
        self.metadata.insert(
            "robot_description_config".to_string(),
            json!({
                "enabled": config.enabled,
                "source": if config.source.trim().is_empty() { "ros" } else { config.source.as_str() },
                "urdf_path": config.urdf_path,
                "base_frame": if config.base_frame.trim().is_empty() { "base_link" } else { config.base_frame.as_str() },
                "ros_param_node": config.ros_param_node,
                "ros_param_name": if config.ros_param_name.trim().is_empty() { "robot_description" } else { config.ros_param_name.as_str() },
                "chunk_size_bytes": config.chunk_size_bytes.clamp(1024, 64000),
                "is_transparent": config.is_transparent,
                "include_visual_meshes": include_visual_meshes,
                "visual_mesh_triangle_budget": config.visual_mesh_triangle_budget.clamp(2000, 500000),
                "body_mesh_mode": body_mesh_mode,
            }),
        );
    }

    pub fn configure_local_body_model(&mut self, robot_model_id: &str, enabled: bool) {
        self.metadata.insert(
            "local_body_model_config".to_string(),
            json!({
                "enabled": enabled,
                "robot_model_id": robot_model_id.trim().to_ascii_lowercase(),
            }),
        );
    }

    pub fn configure_workspace_compass(&mut self, enabled: bool, gateway_port: u16, voice_mode: &str) {
        self.metadata.insert(
            "workspace_compass_config".to_string(),
            json!({
                "enabled": enabled,
                "gateway_port": gateway_port,
                "voice_mode": voice_mode,
                "autonomy": "approve_actions",
                "contract_version": "compass.v1",
            }),
        );
    }

    pub fn configure_workspace_tutorial(&mut self, preset_id: &str) {
        self.metadata.insert(
            "workspace_tutorial_config".to_string(),
            json!({
                "enabled": true,
                "preset_id": preset_id,
            }),
        );
    }

    pub fn add_sensor(&mut self, sensor: Arc<dyn Sensor>) -> Result<(), String> {
        if self.sensors.iter().any(|s| s.name() == sensor.name()) {
            return Err(format!("Sensor with name '{}' already exists", sensor.name()));
        }
        self.sensors.push(sensor);
        Ok(())
    }

    pub fn remove_sensor(&mut self, sensor_name: &str) -> bool {
        let before = self.sensors.len();
        self.sensors.retain(|sensor| sensor.name() != sensor_name);
        before != self.sensors.len()
    }

    pub fn get_sensor(&self, sensor_name: &str) -> Option<SensorRef> {
        self.sensors
            .iter()
            .find(|sensor| sensor.name() == sensor_name)
            .cloned()
    }

    pub fn get_sensors_by_type(&self, sensor_type: SensorType) -> Vec<SensorRef> {
        self.sensors
            .iter()
            .filter(|sensor| sensor.sensor_type() == sensor_type)
            .cloned()
            .collect()
    }

    pub fn get_sensor_count(&self) -> usize {
        self.sensors.len()
    }

    pub fn has_sensors(&self) -> bool {
        !self.sensors.is_empty()
    }

    pub fn create_dataviz(&self, dataviz_name: Option<&str>) -> DataViz {
        let mut dataviz = DataViz::new(
            dataviz_name
                .map(ToString::to_string)
                .unwrap_or_else(|| format!("{}_viz", self.name)),
        );
        for sensor in &self.sensors {
            dataviz.add_sensor_visualization(sensor.as_ref(), &self.name, None);
        }
        dataviz.add_robot_transform(&self.name, "/tf", &self.resolve_tf_frame("base_link"), None);
        dataviz
    }

    pub fn add_path_planning_to_dataviz(
        &self,
        dataviz: &mut DataViz,
        global_path_topic: Option<&str>,
        local_path_topic: Option<&str>,
        trajectory_topic: Option<&str>,
    ) {
        if let Some(topic) = global_path_topic {
            dataviz.add_robot_global_path(&self.name, topic, "map", None);
        }
        if let Some(topic) = local_path_topic {
            dataviz.add_robot_local_path(&self.name, topic, "map", None);
        }
        if let Some(topic) = trajectory_topic {
            dataviz.add_robot_trajectory(&self.name, topic, "map", None);
        }
    }

    pub fn create_full_dataviz(
        &self,
        dataviz_name: Option<&str>,
        global_path_topic: Option<&str>,
        local_path_topic: Option<&str>,
        trajectory_topic: Option<&str>,
    ) -> DataViz {
        let mut dataviz = self.create_dataviz(dataviz_name);
        self.add_path_planning_to_dataviz(
            &mut dataviz,
            global_path_topic,
            local_path_topic,
            trajectory_topic,
        );
        dataviz
    }

    pub fn add_navigation_safety_to_dataviz(&self, dataviz: &mut DataViz) {
        let odom_topic = self.resolve_topic("odom");
        dataviz.add_robot_velocity_data(&self.name, &odom_topic, "map", None);
        dataviz.add_robot_odometry_trail(&self.name, &odom_topic, "map", None);
        dataviz.add_robot_collision_risk(
            &self.name,
            &self.resolve_topic("collision_risk"),
            &self.resolve_tf_frame("base_link"),
            None,
        );
    }

    pub fn register_with_horus(
        &mut self,
        dataviz: Option<DataViz>,
        keep_alive: bool,
        show_dashboard: bool,
        workspace_scale: Option<f64>,
    ) -> (bool, Value) {
        let dataviz = dataviz.unwrap_or_else(|| self.create_dataviz(None));
        let registry = get_robot_registry_client();
        let (success, result) = registry.register_robot(
            self,
            &dataviz,
            10.0,
            keep_alive,
            show_dashboard,
            workspace_scale,
        );
        if success {
            self.add_metadata(
                "horus_topics",
                Value::Array(
                    dataviz
                        .get_enabled_visualizations()
                        .iter()
                        .map(|viz| Value::String(viz.data_source.topic.clone()))
                        .collect(),
                ),
            );
        }
        (success, result)
    }

    pub fn unregister_from_horus(&mut self) -> (bool, Value) {
        let Some(robot_id) = self
            .metadata
            .get("horus_robot_id")
            .and_then(Value::as_str)
            .map(ToString::to_string)
        else {
            return (false, json!({"error": "Robot not registered with HORUS"}));
        };

        let topics: Vec<String> = self
            .metadata
            .get("horus_topics")
            .and_then(Value::as_array)
            .map(|items| {
                items
                    .iter()
                    .filter_map(Value::as_str)
                    .map(ToString::to_string)
                    .collect()
            })
            .unwrap_or_default();

        let registry = get_robot_registry_client();
        let (success, result) = registry.unregister_robot(&robot_id, 5.0);
        if success {
            if !topics.is_empty() {
                let board = get_topic_status_board();
                for topic in &topics {
                    board.on_unsubscribe(topic);
                }
                let monitor = get_topic_monitor();
                monitor.unwatch_topics(&topics, false);
            }
            self.metadata.remove("horus_robot_id");
            self.metadata.remove("horus_color");
            self.metadata.remove("horus_registered");
            self.metadata.remove("horus_topics");
        }
        (success, result)
    }

    pub fn is_registered_with_horus(&self) -> bool {
        self.metadata
            .get("horus_registered")
            .and_then(Value::as_bool)
            .unwrap_or(false)
    }

    pub fn get_horus_id(&self) -> Option<String> {
        self.metadata
            .get("horus_robot_id")
            .and_then(Value::as_str)
            .map(ToString::to_string)
    }

    pub fn get_horus_color(&self) -> Option<String> {
        self.metadata
            .get("horus_color")
            .and_then(Value::as_str)
            .map(ToString::to_string)
    }
}

fn normalize_binding_mode(value: &str) -> &'static str {
    match value.trim().to_ascii_lowercase().as_str() {
        "flat" => "flat",
        _ => "prefixed",
    }
}

fn put_if_some<T: serde::Serialize>(values: &mut serde_json::Map<String, Value>, key: &str, value: Option<T>) {
    if let Some(value) = value {
        values.insert(key.to_string(), json!(value));
    }
}

pub fn register_robots(
    robots: &mut [Robot],
    datavizs: Option<Vec<DataViz>>,
    keep_alive: bool,
    show_dashboard: bool,
    timeout_sec: f64,
    workspace_scale: Option<f64>,
) -> (bool, Value) {
    let registry = get_robot_registry_client();
    registry.register_robots(
        robots,
        datavizs,
        timeout_sec,
        keep_alive,
        show_dashboard,
        workspace_scale,
    )
}

File: cpp/src/robot/robot.cpp
#include "horus/robot/robot.hpp"

#include "horus/bridge/robot_registry.hpp"
#include <algorithm>
#include <cctype>
#include <stdexcept>

namespace horus {
namespace robot {

namespace {
std::optional<std::string> any_to_string(const std::any& value) {
    if (value.type() == typeid(std::string)) {
        return std::any_cast<std::string>(value);
    }
    if (value.type() == typeid(const char*)) {
        return std::string(std::any_cast<const char*>(value));
    }
    return std::nullopt;
}

std::optional<bool> any_to_bool(const std::any& value) {
    if (value.type() == typeid(bool)) {
        return std::any_cast<bool>(value);
    }
    if (value.type() == typeid(int)) {
        return std::any_cast<int>(value) != 0;
    }
    if (value.type() == typeid(double)) {
        return std::any_cast<double>(value) != 0.0;
    }
    return std::nullopt;
}

std::string normalize_binding_mode(std::string value) {
    std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    });
    return value == "flat" ? "flat" : "prefixed";
}

std::string trim_slashes(std::string value) {
    while (!value.empty() && value.front() == '/') {
        value.erase(value.begin());
    }
    while (!value.empty() && value.back() == '/') {
        value.pop_back();
    }
    return value;
}
} // namespace

Robot::Robot(std::string name, core::RobotType robot_type, std::optional<RobotDimensions> dimensions)
    : name_(std::move(name)), robot_type_(robot_type), dimensions_(dimensions) {
    if (name_.empty()) {
        throw std::invalid_argument("Robot name cannot be empty");
    }
    if (dimensions_) {
        if (dimensions_->length < 0.0 || dimensions_->width < 0.0 || dimensions_->height < 0.0) {
            throw std::invalid_argument("Robot dimensions must be non-negative");
        }
    }
}

std::string Robot::get_type_str() const {
    return core::robot_type_to_string(robot_type_);
}

void Robot::add_metadata(const std::string& key, const std::any& value) {
    metadata_[key] = value;
}

std::optional<std::any> Robot::get_metadata(const std::string& key) const {
    const auto it = metadata_.find(key);
    if (it == metadata_.end()) {
        return std::nullopt;
    }
    return it->second;
}

void Robot::configure_ros_binding(
    const std::string& tf_mode,
    const std::string& topic_mode,
    const std::string& base_frame) {
    const auto resolved_tf_mode = normalize_binding_mode(tf_mode);
    const auto resolved_topic_mode = normalize_binding_mode(topic_mode);
    const auto resolved_base_frame = base_frame.empty() ? std::string("base_link") : base_frame;
    metadata_["ros_binding"] = std::map<std::string, std::any>{
        {"logical_name", name_},
        {"tf_mode", resolved_tf_mode},
        {"topic_mode", resolved_topic_mode},
        {"base_frame", resolved_base_frame},
        {"tf_prefix", resolved_tf_mode == "flat" ? std::string("") : name_},
        {"topic_prefix", resolved_topic_mode == "flat" ? std::string("") : "/" + name_},
    };
}

std::map<std::string, std::any> Robot::get_ros_binding() const {
    const auto existing = get_metadata("ros_binding");
    if (existing && existing->type() == typeid(std::map<std::string, std::any>)) {
        return std::any_cast<std::map<std::string, std::any>>(*existing);
    }
    return {
        {"logical_name", name_},
        {"tf_mode", std::string("prefixed")},
        {"topic_mode", std::string("prefixed")},
        {"base_frame", std::string("base_link")},
        {"tf_prefix", name_},
        {"topic_prefix", "/" + name_},
    };
}

std::string Robot::resolve_topic(const std::string& topic_name) const {
    if (topic_name.empty()) {
        return "";
    }
    if (topic_name.front() == '/') {
        return topic_name;
    }
    const auto binding = get_ros_binding();
    std::string prefix;
    const auto it = binding.find("topic_prefix");
    if (it != binding.end()) {
        prefix = any_to_string(it->second).value_or("");
    }
    if (prefix.empty()) {
        return "/" + topic_name;
    }
    while (!prefix.empty() && prefix.back() == '/') {
        prefix.pop_back();
    }
    return prefix + "/" + topic_name;
}

std::string Robot::resolve_tf_frame(const std::string& frame_id) const {
    auto frame = trim_slashes(frame_id);
    if (frame.empty()) {
        return "";
    }
    const auto binding = get_ros_binding();
    std::string prefix;
    const auto it = binding.find("tf_prefix");
    if (it != binding.end()) {
        prefix = trim_slashes(any_to_string(it->second).value_or(""));
    }
    if (prefix.empty() || frame.rfind(prefix + "/", 0) == 0) {
        return frame;
    }
    return prefix + "/" + frame;
}

void Robot::configure_robot_manager(bool status, bool data_viz, bool teleop, bool tasks) {
    metadata_["robot_manager_config"] = std::map<std::string, std::any>{
        {"enabled", true},
        {"prefab_asset_path", std::string("Assets/Prefabs/UI/RobotManager.prefab")},
        {"prefab_resource_path", std::string("")},
        {"sections",
         std::map<std::string, std::any>{
             {"status", status},
             {"data_viz", data_viz},
             {"teleop", teleop},
             {"tasks", tasks},
         }},
    };
}

void Robot::configure_local_body_model(const std::string& robot_model_id, bool enabled) {
    metadata_["local_body_model_config"] = std::map<std::string, std::any>{
        {"enabled", enabled},
        {"robot_model_id", robot_model_id},
    };
}

void Robot::configure_workspace_compass(bool enabled, int gateway_port, const std::string& voice_mode) {
    metadata_["workspace_compass_config"] = std::map<std::string, std::any>{
        {"enabled", enabled},
        {"gateway_port", gateway_port},
        {"voice_mode", voice_mode},
        {"autonomy", std::string("approve_actions")},
        {"contract_version", std::string("compass.v1")},
    };
}

void Robot::configure_workspace_tutorial(const std::string& preset_id, bool enabled) {
    metadata_["workspace_tutorial_config"] = std::map<std::string, std::any>{
        {"enabled", enabled},
        {"preset_id", preset_id},
    };
}

void Robot::add_sensor(const std::shared_ptr<Sensor>& sensor) {
    if (!sensor) {
        throw std::invalid_argument("Sensor cannot be null");
    }
    const auto duplicate = std::find_if(sensors_.begin(), sensors_.end(), [&](const auto& existing) {
        return existing->get_name() == sensor->get_name();
    });
    if (duplicate != sensors_.end()) {
        throw std::invalid_argument("Sensor with name '" + sensor->get_name() + "' already exists");
    }
    sensors_.push_back(sensor);
}

bool Robot::remove_sensor(const std::string& sensor_name) {
    const auto before = sensors_.size();
    sensors_.erase(
        std::remove_if(
            sensors_.begin(),
            sensors_.end(),
            [&](const auto& sensor) { return sensor->get_name() == sensor_name; }),
        sensors_.end());
    return before != sensors_.size();
}

std::shared_ptr<Sensor> Robot::get_sensor(const std::string& sensor_name) const {
    const auto it = std::find_if(sensors_.begin(), sensors_.end(), [&](const auto& sensor) {
        return sensor->get_name() == sensor_name;
    });
    if (it == sensors_.end()) {
        return nullptr;
    }
    return *it;
}

std::vector<std::shared_ptr<Sensor>> Robot::get_sensors_by_type(core::SensorType type) const {
    std::vector<std::shared_ptr<Sensor>> result;
    for (const auto& sensor : sensors_) {
        if (sensor->get_sensor_type() == type) {
            result.push_back(sensor);
        }
    }
    return result;
}

std::shared_ptr<dataviz::DataViz> Robot::create_dataviz(const std::string& dataviz_name) const {
    const auto name = dataviz_name.empty() ? (name_ + "_viz") : dataviz_name;
    auto dataviz = std::make_shared<dataviz::DataViz>(name);
    for (const auto& sensor : sensors_) {
        dataviz->add_sensor_visualization(sensor, name_);
    }
    dataviz->add_robot_transform(name_, "/tf", resolve_tf_frame("base_link"));
    return dataviz;
}

void Robot::add_navigation_safety_to_dataviz(dataviz::DataViz& dataviz) const {
    const auto odom_topic = resolve_topic("odom");
    dataviz.add_robot_velocity_data(name_, odom_topic, "map");
    dataviz.add_robot_odometry_trail(name_, odom_topic, "map");
    dataviz.add_robot_collision_risk(
        name_,
        resolve_topic("collision_risk"),
        resolve_tf_frame("base_link"));
}

void Robot::add_path_planning_to_dataviz(
    dataviz::DataViz& dataviz,
    const std::optional<std::string>& global_path_topic,
    const std::optional<std::string>& local_path_topic,
    const std::optional<std::string>& trajectory_topic) const {
    if (global_path_topic) {
        dataviz.add_robot_global_path(name_, *global_path_topic, "map");
    }
    if (local_path_topic) {
        dataviz.add_robot_local_path(name_, *local_path_topic, "map");
    }
    if (trajectory_topic) {
        dataviz.add_robot_trajectory(name_, *trajectory_topic, "map");
    }
}

std::shared_ptr<dataviz::DataViz> Robot::create_full_dataviz(
    const std::string& dataviz_name,
    const std::optional<std::string>& global_path_topic,
    const std::optional<std::string>& local_path_topic,
    const std::optional<std::string>& trajectory_topic) const {
    auto dataviz = create_dataviz(dataviz_name);
    add_path_planning_to_dataviz(*dataviz, global_path_topic, local_path_topic, trajectory_topic);
    return dataviz;
}

std::pair<bool, std::map<std::string, std::any>> Robot::register_with_horus(
    std::shared_ptr<dataviz::DataViz> dataviz,
    bool keep_alive,
    bool show_dashboard,
    std::optional<double> workspace_scale) {
    if (!dataviz) {
        dataviz = create_dataviz();
    }
    auto& registry = bridge::get_robot_registry_client();
    auto result =
        registry.register_robot(*this, *dataviz, 10.0, keep_alive, show_dashboard, workspace_scale);
    if (result.first) {
        metadata_["horus_registered"] = true;
    }
    return result;
}

std::pair<bool, std::map<std::string, std::any>> Robot::unregister_from_horus() {
    auto id = get_horus_id();
    if (!id) {
        return {false, {{"error", std::string("Robot not registered with HORUS")}}};
    }
    auto& registry = bridge::get_robot_registry_client();
    auto result = registry.unregister_robot(*id, 5.0);
    if (result.first) {
        metadata_.erase("horus_robot_id");
        metadata_.erase("horus_color");
        metadata_.erase("horus_registered");
        metadata_.erase("horus_topics");
    }
    return result;
}

bool Robot::is_registered_with_horus() const {
    auto value = get_metadata("horus_registered");
    if (!value) {
        return false;
    }
    auto bool_value = any_to_bool(*value);
    return bool_value.value_or(false);
}

std::optional<std::string> Robot::get_horus_id() const {
    auto value = get_metadata("horus_robot_id");
    if (!value) {
        return std::nullopt;
    }
    return any_to_string(*value);
}

std::optional<std::string> Robot::get_horus_color() const {
    auto value = get_metadata("horus_color");
    if (!value) {
        return std::nullopt;
    }
    return any_to_string(*value);
}

std::pair<bool, std::map<std::string, std::any>> register_robots(
    std::vector<Robot*>& robots,
    std::vector<dataviz::DataViz> datavizs,
    double timeout_sec,
    bool keep_alive,
    bool show_dashboard,
    std::optional<double> workspace_scale) {
    auto& registry = bridge::get_robot_registry_client();
    return registry.register_robots(
        robots,
        std::move(datavizs),
        timeout_sec,
        keep_alive,
        show_dashboard,
        workspace_scale);
}

} // namespace robot
} // namespace horus
